### DISCUSSION TITLE
Help fixing strange syntax highlighting with hocon grammar

### DISCUSSION THREAD

**Discussion by LaurenceWarne** [github-linguist/linguist#6253] [Q&A] [answered]
Hi, I've noticed the syntax highlighting of hocon on Github appears to be broken, but I can't reproduce the strange highlighting when I use the same grammar in an editor:
Github            |  Using Grammar from Editor (Atom)
:-------------------------:|:-------------------------:
![hocon-weird](https://user-images.githubusercontent.com/17688577/212745898-ceba0d6a-9a11-4d2b-a6fc-10d37bd8343e.png) |  ![hocon-correct](https://user-images.githubusercontent.com/17688577/212746340-42805412-c6c7-47a9-bad1-407b9324fa18.png)

An even smaller case is:
```hocon
foo = "bar"
```
( the string should be highlighted)

I was wondering if anyone had any thoughts on how this can be fixed, I'm more than happy to submit/suggest PRs upstream to the grammar repo if necessary.

Unsure if this is relevant, but using https://umanghome.github.io/NovaLightshow/ with the grammar gives the error 
```
TypeError: Expected `string` for `scope`, got `undefined
````
Though converting the grammar from cson to json gives correct highlighting.

For reference the grammar is located here: https://github.com/jacobwgillespie/language-hocon/blob/master/grammars/hocon.cson

Thanks!

**Comment by lildude** [comment]
The difference in behaviour is usually down to the use of Oniguruma-specific regexes. Atom uses Oniguruma (as does NovaLightShow) for the regex engine whilst GitHub uses PCRE for performance reasons. There's also the possibility there's an error which GitHub is less tolerant of than Atom.

I'm rubbish with grammars, but maybe @Alhadis or @Nixinova can spot the problem.

**Comment by Nixinova** [comment]
The highlighting issues on NovaLightshow's end are from https://github.com/fabiospampinato/cson2json/issues/1. I should look into switching packages to avoid that.

> The difference in behaviour is usually down to the use of Oniguruma-specific regexes. 

Hm, none of the regexes used have anything non agnostic to them.
Since the grammar works fine when converted to JSON I don't think it would be an issue with the grammar itself? At least it means that I wouldn't know how to debug the grammar since it, well, works.

**Comment by Alhadis** [comment, marked as answer]
I suspect this may have something to do with the malformed rule at [lines 43-49](https://github.com/jacobwgillespie/language-hocon/blob/99e233e959529ef304d39acb8414ab3fb21ba558/grammars/hocon.cson#L43-L49):

~~~cson
    match: '(?:^[ \t]*([\\w-]+)\\s*?({|=|:))'
    captures:
      '1':
        'name': 'entity.name.tag.hocon'
      '2':
        'name': 'punctuation.separator.key-value.hocon'
    end: '(\\=|{)'
~~~

You've specified an `end` pattern without an accompanying `begin` pattern, and done so whilst including a `match` field (which is mutually exclusive to a `begin`/`end` pair). GitHub and Atom probably differ in their handling of malformed grammar rules, accounting for the discrepancy that you've witnessed.

**Reply by RedCMD** [reply]
If an `end` exists
TextMate 2.0 will treat the `match` as a `begin`
even if a `begin` exists, it will still use the `match`!!

so this a bug with Github

**Comment by LaurenceWarne** [comment]
Thanks for all the input
> You've specified an end pattern without an accompanying begin pattern, and done so whilst including a match field (which is mutually exclusive to a begin/end pair). GitHub and Atom probably differ in their handling of malformed grammar rules, accounting for the discrepancy that you've witnessed.

:+1:  I've opened a PR for it here: https://github.com/jacobwgillespie/language-hocon/pull/7 (comments welcome), let's see what the maintainer says.

As an aside, is there a way to be sure the change will work prior to a new release?

**Reply by LaurenceWarne** [reply]
Update: maintainer has merged the PR :tada:

**Comment by LaurenceWarne** [comment]
Looks to have worked:
![hocon-fixed](https://user-images.githubusercontent.com/17688577/230653709-ccfe8c1a-1cc0-4f60-aeb4-9eb2e1644522.png)
Thanks again!
